aten/src/ATen/native/cuda/AmpKernels.cu

```diff
@@ -0,0 +1,93 @@
+#define _USE_MATH_DEFINES
+
+#include <math.h>
+
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cuda/Loops.cuh>
+
+
+namespace at {
+namespace native {
+
+// These functions have no CPU variants, so there's no need to bother with a .cpp file in native/, and
+// also no need to impose the DECLARE/DEFINE/REGISTER_DISPATCH layer of indirection.
+
+Tensor & _amp_unscale_inf_check_cuda(Tensor & scaled_grad,
+                                     const Tensor & current_scale,
+                                     const Tensor & found_inf)
+{
+  // Do I really need all these checks?
+  TORCH_CHECK(scaled_grad.is_cuda(), "scaled_grad must be a CUDA tensor.");
+  TORCH_CHECK(current_scale.is_cuda(), "current_scale must be a CUDA tensor.");
+  TORCH_CHECK(found_inf.is_cuda(), "found_inf must be a CUDA tensor.");
+  TORCH_CHECK(current_scale.numel() == 1, "current_scale must be a 1-element tensor.");
+  TORCH_CHECK(found_inf.numel() == 1, "found_inf must be a 1-element tensor.");
+  TORCH_CHECK(current_scale.scalar_type() == at::ScalarType::Float, "current_scale must be a float tensor.");
+  TORCH_CHECK(found_inf.scalar_type() == at::ScalarType::Float, "found_inf must be a float tensor.");
+
+  auto rscale = current_scale.to(at::ScalarType::Double).reciprocal();
+
+  // Act on scaled_grad in place.
+  auto iter = TensorIterator::unary_op(scaled_grad, scaled_grad);
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+    iter.dtype(),
+    "_amp_unscale_inf_check_kernel",
+    [&] {
+      float* found_inf_ptr = found_inf.data_ptr<float>();
+      double* rscale_ptr = rscale.data_ptr<double>();
+
+      gpu_kernel(iter, [found_inf_ptr, rscale_ptr] GPU_LAMBDA(scalar_t val) -> scalar_t {
+          float fval = static_cast<float>(val);
+          if(!std::isfinite(fval))
+            *found_inf_ptr = 1.f;
+          return static_cast<scalar_t>(fval*(*rscale_ptr));
```

Comment:
You're multiplying a float by a double here and then casting the result to the scalar type? Can you elaborate on the choice of types?

Reply:
I found that the default operator/ applied to float values gave lousy accuracy.  My tests (which compare param values after several steps on a dummy model with scaling+unscaling versus several steps on the same model with no scaling) did not pass to within `torch.allclose`'s default tolerances ([rtol=1e-5, atol=1e-8](https://pytorch.org/docs/stable/torch.html#torch.allclose)).

The above pattern does all the math in FP64, which improved accuracy and allowed all my tests to pass to within `torch.allclose`'s default FP32 tolerances (except for one of the gradient clipping tests, which required reducing the absolute tolerance from 1e-8 to 1e-7, but we can discuss that separately).  Volta and Turing have good fp64 multiply throughput, so the kernel was still bandwidth bound on my machine.

However, I just realized type conversion to/from fp64 may be a bottleneck on Turing (buried in a footnote, the cuda C guide says Turing fp64 conversions have throughput of just [2 ops per SM per cycle](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#fnsrc_8) :grimacing:   I can play around with this to find a pattern that uses pure FP32 while maintaining accuracy.